# Practicing SQL Queries

## 1

In [7]:
from sqlalchemy import create_engine
import pandas as pd

In [8]:
import sqlalchemy
import pymysql
import IPython
print("All libraries imported successfully!")

All libraries imported successfully!


In [9]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [10]:
%sql mysql+pymysql://student:guest@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft

### 1a) Show all tables in the database.

In [11]:
%%sql
SHOW TABLES;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
37 rows affected.


[('accused',),
 ('accused_family',),
 ('appeal',),
 ('calendarcustom',),
 ('case_person',),
 ('commission',),
 ('complaint',),
 ('confession',),
 ('counterstrategy',),
 ('demonicpact',),
 ('denunciation',),
 ('devilappearance',),
 ('elf_fairyelements',),
 ('imprisonment',),
 ('linkedtrial',),
 ('malice',),
 ('mentionedaswitch',),
 ('movestohla',),
 ('musicalinstrument',),
 ('ordeal',),
 ('othercharges',),
 ('othernamedwitch',),
 ('person',),
 ('prevcommission',),
 ('propertydamage',),
 ('ref_parish',),
 ('reference',),
 ('religiousmotif',),
 ('ritualobject',),
 ('shapechanging',),
 ('source',),
 ('torture',),
 ('trial',),
 ('trial_person',),
 ('weathermodification',),
 ('whitemagic',),
 ('witchesmeetingplace',)]

### 1b) Show the first three entries in the table accused.

In [13]:
%%sql
SELECT * FROM accused LIMIT 3;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
3 rows affected.


[('A/EGD/10', 'EGD', 10, 'Mareon', 'Quheitt', 'Marion', 'White', None, None, None, 'Female', None, 0, 0, 'Sammuelston', 'P/JO/3539', 'Haddington', 'Haddington', None, None, None, None, None, None, None, None, None, 'SMD', '2001-05-15T11:06:51', 'jhm', '2002-08-09T11:40:51'),
 ('A/EGD/100', 'EGD', 100, 'Thom', 'Cockburn', 'Thomas', 'Cockburn', None, None, None, 'Male', None, 0, 0, None, None, None, 'Haddington', None, None, None, None, None, None, None, None, None, 'SMD', '2001-05-15T11:06:51', 'jhm', '2002-10-02T10:32:30'),
 ('A/EGD/1000', 'EGD', 1000, 'Christian', 'Aitkenhead', 'Christine', 'Aikenhead', None, None, None, 'Female', None, 0, 0, 'Rottinraw', None, None, 'Dumfries', None, None, None, None, None, 'Married', None, None, None, 'SMD', '2001-05-15T11:06:51', 'jhm', '2002-10-01T10:48:12')]

### 1c) How many people are included in the accused table?

In [14]:
%%sql
SELECT COUNT(*) AS total_accused FROM accused;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3219,)]

### 1d) Looks like the age is missing for some observations. Count the number of non-missing values for age in the data.

In [15]:
%%sql
SELECT COUNT(age) AS age_count FROM accused WHERE age IS NOT NULL;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(166,)]

### 1e) Show a list of unique occupations

In [16]:
%%sql
SELECT DISTINCT occupation FROM accused;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
32 rows affected.


[(None,),
 ('Servant',),
 ('Vagabond',),
 ('Weaver',),
 ('Midwife',),
 ('Tailor',),
 ('Messenger',),
 ('Brewster',),
 ('Smith',),
 ('Minister',),
 ('Farmer',),
 ('Mealmaker',),
 ('Tasker',),
 ('Maltman',),
 ('Miller',),
 ('Healer',),
 ('Loadman',),
 ('Blacksmith',),
 ('Shop-keeper',),
 ('Stabler',),
 ('Nurse',),
 ('Collier',),
 ('School teacher',),
 ('Slaterer',),
 ('Workman',),
 ('School Master',),
 ('Merchant',),
 ('Fisherman',),
 ('Henwife',),
 ('Cook',),
 ('Sailor',),
 ('Creelman',)]

### 1f) What proportion of accused are female?

In [17]:
%%sql
SELECT COUNT(*) AS female_count FROM accused WHERE LOWER(sex) = 'female';

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(2702,)]

In [18]:
%%sql
SELECT COUNT(*) AS total_count FROM accused;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3219,)]

In [20]:
num_female = %sql SELECT COUNT(*) AS female_count FROM accused WHERE LOWER(sex) = 'female';
total = %sql SELECT COUNT(*) AS total_count FROM accused;

# Convert the ResultSet to integers (access the first element of the first row)
female_count = int(num_female[0][0])
total_count = int(total[0][0])

# Calculate the proportion
female_proportion = round(female_count / total_count, 2)
print(f"Proportion of accused that are female: {female_proportion}")

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
Proportion of accused that are female: 0.84


## 2  Seeing the Devil

### 2a) List the unique devil_types in the data.

In [22]:
%%sql
SELECT DISTINCT devil_type FROM devilappearance;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
16 rows affected.


[('Male',),
 ('Female Fairy',),
 ('Male Fairy',),
 ('Animal Devil',),
 ('Spirit',),
 ('Ghost',),
 ('Other Demon',),
 ('Female',),
 (None,),
 ('Fairy',),
 ('Unspecified Devil',),
 ('Insect Devil',),
 ('Baby',),
 ('Human Devil',),
 ('Child Devil',),
 ('Inanimate Object Devil',)]

### 2b) There is also a little description of the type of the devil sighting in the devil_text column. How many of the sightings mention the word "black" in the description? 

In [23]:
%%sql
SELECT COUNT(*) AS black_mention_count
FROM devilappearance
WHERE LOWER(devil_text) LIKE '%black%';

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(121,)]

### 2c) What proportion of the devils (in devil_type) are male?

In [27]:
%%sql
SELECT (COUNT(*) / (SELECT COUNT(*) FROM devilappearance)) AS male_proportion
FROM devilappearance
WHERE devil_type = 'male';

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(Decimal('0.6313'),)]

## 3

### a)

In [34]:
%%sql
SELECT 
    MAX(male_accusers), 
    MAX(female_accusers), 
    AVG(male_accusers), 
    AVG(female_accusers)
FROM trial;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(48.0, 27.0, 0.4181988158304768, 0.2559226932668329)]

### b)

In [38]:
%%sql
SELECT 
    sentence AS sentence_type,    
    COUNT(*) AS sentence_count    
FROM trial
WHERE sentence IS NOT NULL        
GROUP BY sentence                
ORDER BY sentence_count DESC;  

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
10 rows affected.


[('Execution', 205),
 ('Released', 52),
 ('Banishment', 27),
 ('Declared Fugitive', 11),
 ('Excommunicated', 6),
 ('Put to the horn', 2),
 ('Hang', 1),
 ('Branded', 1),
 ('Prison', 1),
 ('Public Humiliation', 1)]

### c

In [39]:
%%sql
SELECT 
    verdict AS verdict_type,                        
    AVG(male_accusers + female_accusers) AS avg_accusers 
FROM trial
WHERE verdict IS NOT NULL                          
GROUP BY verdict                                   
ORDER BY avg_accusers DESC;     

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
4 rows affected.


[('Guilty', 5.804255319148936),
 ('Not Guilty', 4.558139534883721),
 ('Not Proven', 1.0),
 ('Half Guilty', 0.2857142857142857)]

## Yes, the number of accusers appears to matter for the verdict. The table below shows the average number of accusers for each verdict type:

## 4a)

In [42]:
%%sql
SELECT 
    COUNT(*) AS total_trials
FROM trial;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3211,)]

In [44]:
%%sql
SELECT 
    COUNT(DISTINCT confessionref) AS total_confessions
FROM confession;


 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(941,)]

In [46]:
%%sql
SELECT COUNT(DISTINCT t.trialref) AS num_trials_with_confessions,
       COUNT(DISTINCT t.trialref) * 1.0 / 3211 AS share_with_confessions
FROM trial AS t
LEFT JOIN confession AS c
ON t.trialref = c.trialref # join on the basis of trialref
WHERE c.confessionref IS NOT NULL

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(756, Decimal('0.23544'))]

In [47]:
%%sql
SELECT 
    (SELECT COUNT(DISTINCT trialref) FROM trial) AS num_total_trials,
    (SELECT COUNT(DISTINCT confessionref) FROM confession) AS num_confessions,
    (SELECT COUNT(DISTINCT t.trialref) 
     FROM trial AS t
     LEFT JOIN confession AS c
     ON t.trialref = c.trialref
     WHERE c.confessionref IS NOT NULL) AS num_trials_with_confessions,
    (SELECT COUNT(DISTINCT t.trialref) * 1.0 / num_total_trials
     FROM trial AS t
     LEFT JOIN confession AS c
     ON t.trialref = c.trialref
     WHERE c.confessionref IS NOT NULL) AS share_with_confessions

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3211, 941, 756, Decimal('0.23544'))]

### b

In [48]:
%sql DESCRIBE torture

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
12 rows affected.


[('tortureref', 'text', 'YES', '', None, ''),
 ('torturesystemid', 'text', 'YES', '', None, ''),
 ('tortureid', 'bigint', 'YES', '', None, ''),
 ('trialref', 'text', 'YES', '', None, ''),
 ('tortureexists', 'tinyint(1)', 'YES', '', None, ''),
 ('torturedate', 'text', 'YES', '', None, ''),
 ('torturedate_as_date', 'text', 'YES', '', None, ''),
 ('torturetype', 'text', 'YES', '', None, ''),
 ('createdby', 'text', 'YES', '', None, ''),
 ('createdate', 'text', 'YES', '', None, ''),
 ('lastupdatedby', 'text', 'YES', '', None, ''),
 ('lastupdatedon', 'text', 'YES', '', None, '')]

In [49]:
%sql SELECT COUNT(*), COUNT(DISTINCT tortureref) FROM torture

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(110, 110)]

In [50]:
%%sql
WITH 
    trials_with_torture AS (
        SELECT DISTINCT trialref
        FROM torture
    ),
    trials_with_confessions AS (
        SELECT DISTINCT trialref
        FROM confession
    )
SELECT
    COUNT(DISTINCT tt.trialref) AS num_torture_trials,
    COUNT(DISTINCT tc.trialref) AS num_torture_confession_trials,
    COUNT(DISTINCT tc.trialref) * 1.0 / COUNT(DISTINCT tt.trialref) AS share_confessions_with_torture
FROM trials_with_torture AS tt
LEFT JOIN trials_with_confessions AS tc
ON tt.trialref = tc.trialref

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(52, 22, Decimal('0.42308'))]

In [51]:
%%sql
WITH
    trials_with_confessions AS (
        SELECT DISTINCT trialref
        FROM confession
    )
SELECT 
    COUNT(DISTINCT t.trialref) AS num_no_torture_trials,
    COUNT(DISTINCT ntc.trialref) AS num_no_torture_confession_trials,
    COUNT(DISTINCT ntc.trialref) * 1.0 / COUNT(DISTINCT t.trialref) AS share_confessions_without_torture
FROM trial AS t
LEFT JOIN trials_with_confessions AS ntc
ON t.trialref = ntc.trialref
WHERE t.trialref NOT IN (SELECT trialref FROM torture)

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3159, 734, Decimal('0.23235'))]

## comparing the value, we notice 42% is higher than 23%, with records of torture is higher than trials without records. So, torture may have significantly influenced confessions during trials